Source/Core/DolphinQt2/MainWindow.cpp

```diff
@@ -86,6 +91,35 @@
 #include "UICommon/X11Utils.h"
 #endif
 
+#if defined(__unix__) || defined(__unix) || defined(__APPLE__)
+void MainWindow::OnSignal()
+{
+  static bool once = false;
```

Comment:
This is not necessary because the signal handler will be reset to the default one automatically.

Reply:
It will cause random segfaults without it.

Reply:
SignalDaemon::HandleInterrupt will never get called, and so OnSignal neither, as far as I can tell. Could you explain how this causes segfaults and where it's segfaulting?

Reply:
Why wouldn't it get called? It's registered as the signal handler in ``MainWindow::InstallSignalHandler``. The segfault is a bit hard to debug since gdb seems to not really trigger ``SIGINT`` like it is when running it normally.

Reply:
SA_RESETHAND is set for the handler flags in `InstallSignalHandler`.

The manpage for sigaction(2) says:

           SA_RESETHAND
                  Restore the signal action to the default upon entry to the
                  signal handler.  This flag is meaningful only when estab‐
                  lishing a signal handler.  SA_ONESHOT is an obsolete, non‐
                  standard synonym for this flag.

Reply:
You said
> SignalDaemon::HandleInterrupt will **never** get called

If you wanna know why it gets called more than once, I don't know. Maybe ``QSocketNotifier::activated`` fires twice? When printing something every time ``OnSignal`` gets called I get the following output.

```
A signal was received. A second signal will force Dolphin to stop.
OnSignal called
```

Reply:
I meant it will never get called *after it's reset*. After re-reading my comment, I guess I didn't make that clear enough, sorry.

Reply:
It *shouldn't* it *does* though.

Reply:
Oh, I think I know what's going on. Maybe the activated signal fires continuously while the socket can be read from.

So the proper fix would be to read from the socket or disable the watcher to stop this from being ever called in the first place.